File: examples/75-media-ole-api/main.go
// examples/75-media-ole-api demonstrates media (audio/video) and OLE object embedding.
//
// Uses the PresentationEditor to insert:
//   - Audio from bytes (tiny WAV) with an icon
//   - Video from file if a fixture exists, otherwise logs a skip notice
//   - OLE object from bytes if a fixture exists, otherwise logs a skip notice
//
// Real fixture files can be placed in examples/assets/75/:
//   - sample.mp4  (video)
//   - poster.png  (video poster / OLE icon)
//   - sample_ole.bin (OLE binary)
//
// Run with: go run ./examples/75-media-ole-api/main.go
package main

import (
	"encoding/base64"
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/editor"
)

const (
	outputDir  = "examples/output"
	outputFile = "75_media_ole_api.pptx"
	assetsDir  = "examples/assets/75"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	tmpDir, err := os.MkdirTemp("", "gopptx-media-ole-*")
	if err != nil {
		return fmt.Errorf("create temp dir: %w", err)
	}
	defer os.RemoveAll(tmpDir)

	basePath := filepath.Join(tmpDir, "base.pptx")

	baseSlides := []pptx.SlideContent{
		pptx.NewSlide("Media & OLE Embedding").
			AddBullet("Audio from bytes – embedded using AddAudioWithIcon.").
			AddBullet("Video from file – requires sample.mp4 + poster.png fixture.").
			AddBullet("OLE object – requires sample_ole.bin + poster.png fixture."),
	}
	if err := pptx.WriteFile(basePath, "Media & OLE API Demo", baseSlides); err != nil {
		return fmt.Errorf("write base deck: %w", err)
	}

	ed, err := pptx.OpenPresentationEditor(basePath)
	if err != nil {
		return fmt.Errorf("open editor: %w", err)
	}
	defer ed.Close()

	// --- 1. Audio from bytes (always works – no fixture needed) ---
	audioOpts := editor.NewAutoPlayAudioPlaybackOptions().
		WithLoop(false).
		WithPlayAcrossSlides(false).
		WithVolume(80).
		WithAltText("Embedded WAV audio clip")

	if _, err := ed.AddAudioWithPlaybackOptions(
		0,
		tinyWAV(),
		"audio/wav",
		audioOpts,
		4500000, // x EMU
		1500000, // y EMU
		1800000, // cx EMU
		900000,  // cy EMU
	); err != nil {
		return fmt.Errorf("add audio: %w", err)
	}
	log.Println("Inserted WAV audio from bytes.")

	// --- 2. Video from file (optional fixture) ---
	videoPath := filepath.Join(assetsDir, "sample.mp4")
	posterPath := filepath.Join(assetsDir, "poster.png")
	if fileExists(videoPath) && fileExists(posterPath) {
		videoOpts := editor.NewAutoPlayVideoPlaybackOptions().
			WithLoop(false).
			WithMuted(true).
			WithVolume(0).
			WithAltText("Sample video clip")
		if _, err := ed.AddVideoFromFileWithPlaybackOptions(
			0, videoPath, posterPath, "video/mp4",
			videoOpts,
			600000, 1400000, 3600000, 2100000,
		); err != nil {
			return fmt.Errorf("add video: %w", err)
		}
		log.Printf("Inserted video from fixture: %s\n", videoPath)
	} else {
		log.Printf("Video fixture not found (%s + %s); skipping.\n", videoPath, posterPath)
	}

	// --- 3. OLE object from bytes (optional fixture) ---
	olePath := filepath.Join(assetsDir, "sample_ole.bin")
	if fileExists(olePath) && fileExists(posterPath) {
		oleData, readErr := os.ReadFile(olePath)
		if readErr != nil {
			return fmt.Errorf("read OLE fixture: %w", readErr)
		}
		if _, err := ed.AddOLEObject(
			0,
			oleData,
			tinyPNG(),
			"Excel.Sheet.12",
			4500000, 2800000, 2100000, 1400000,
		); err != nil {
			return fmt.Errorf("add OLE object: %w", err)
		}
		log.Printf("Inserted OLE object from fixture: %s\n", olePath)
	} else {
		log.Printf("OLE fixture not found (%s); skipping.\n", olePath)
	}

	outPath := filepath.Join(outputDir, outputFile)
	if err := ed.Save(outPath); err != nil {
		return fmt.Errorf("save output: %w", err)
	}

	log.Printf("Generated %s\n", outPath)
	return nil
}

func tinyWAV() []byte {
	return []byte{
		'R', 'I', 'F', 'F',
		0x25, 0x00, 0x00, 0x00,
		'W', 'A', 'V', 'E',
		'f', 'm', 't', ' ',
		0x10, 0x00, 0x00, 0x00,
		0x01, 0x00,
		0x01, 0x00,
		0x40, 0x1F, 0x00, 0x00,
		0x40, 0x1F, 0x00, 0x00,
		0x01, 0x00,
		0x08, 0x00,
		'd', 'a', 't', 'a',
		0x01, 0x00, 0x00, 0x00,
		0x80,
	}
}

func tinyPNG() []byte {
	const b64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8Xw8AAoMBgA3FoXwAAAAASUVORK5CYII="
	data, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		panic(err)
	}
	return data
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

File: examples/75-media-ole-api/main.py
"""Demonstrate media embedding and document OLE embedding patterns.

This example demonstrates:
- Embedding audio from raw bytes with add_audio()
- Embedding video from a file path with add_video() (requires fixture)
- Reference slides covering OLE embedding concepts
- Graceful skip when optional fixture files are not present
"""

from __future__ import annotations

import tempfile
from pathlib import Path

from gopptx import Presentation
from gopptx.presentation.slides import SlideLayoutType
from gopptx.schemas import Inches

# Minimal valid WAV (PCM mono, 8-bit, 8000 Hz, 1 sample).
_TINY_WAV = bytes([
    ord("R"),
    ord("I"),
    ord("F"),
    ord("F"),
    0x25,
    0x00,
    0x00,
    0x00,
    ord("W"),
    ord("A"),
    ord("V"),
    ord("E"),
    ord("f"),
    ord("m"),
    ord("t"),
    ord(" "),
    0x10,
    0x00,
    0x00,
    0x00,
    0x01,
    0x00,
    0x01,
    0x00,
    0x40,
    0x1F,
    0x00,
    0x00,
    0x40,
    0x1F,
    0x00,
    0x00,
    0x01,
    0x00,
    0x08,
    0x00,
    ord("d"),
    ord("a"),
    ord("t"),
    ord("a"),
    0x01,
    0x00,
    0x00,
    0x00,
    0x80,
])

ASSETS_DIR = Path("examples/assets/75")


def _add_overview_slide(prs: Presentation) -> int:
    """Overview slide listing media embedding methods."""
    return prs.add_bullet_slide(
        "Media & OLE Embedding",
        [
            "Audio from bytes  - add_audio(slide_idx, path, x, y, w, h)",
            "Video from file   - add_video(slide_idx, path, x, y, w, h)",
            "OLE objects       - available via PresentationEditor (Go API)",
            "Fixture files in  examples/assets/75/ for video/OLE demos",
        ],
    )


def _add_audio_slide(prs: Presentation, wav_path: Path) -> None:
    """Embed audio from a temp WAV file."""
    idx = prs.add_slide("Audio Embedding", layout=SlideLayoutType.TITLE_ONLY)
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(1.2),
        Inches(9),
        Inches(0.6),
        text=f"Audio source: {wav_path.name}  (tiny WAV, 1 sample)",
    )
    prs.add_audio(
        idx,
        str(wav_path),
        Inches(4.5),
        Inches(1.5),
        Inches(1.8),
        Inches(0.9),
    )
    print(f"  Inserted WAV audio from bytes (via temp file: {wav_path})")


def _add_video_slide(prs: Presentation) -> None:
    """Embed video if fixture files exist, otherwise log a skip notice."""
    video_path = ASSETS_DIR / "sample.mp4"
    if not video_path.exists():
        prs.add_bullet_slide(
            "Video Embedding (fixture not found)",
            [
                f"Expected fixture: {video_path}",
                "Place sample.mp4 in examples/assets/75/ to enable this demo.",
                "add_video(slide_idx, path, x, y, w, h)",
            ],
        )
        print(f"  Video fixture not found ({video_path}); skipping.")
        return

    idx = prs.add_slide("Video Embedding", layout=SlideLayoutType.TITLE_ONLY)
    prs.add_video(
        idx,
        str(video_path),
        Inches(0.6),
        Inches(1.4),
        Inches(3.6),
        Inches(2.1),
    )
    print(f"  Inserted video from fixture: {video_path}")


def _add_ole_reference_slide(prs: Presentation) -> None:
    """Reference slide for OLE object embedding concepts."""
    prs.add_bullet_slide(
        "OLE Object Embedding (Go API)",
        [
            "OLE embedding requires the PresentationEditor API in Go.",
            "ed.AddOLEObject(slideIdx, data, icon, progID, x, y, cx, cy)",
            "Supported ProgIDs: Excel.Sheet.12, Word.Document.12, etc.",
            "Place sample_ole.bin + poster.png in examples/assets/75/ for Go demo.",
        ],
    )


def main() -> None:
    """Create presentation demonstrating media and OLE embedding."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    output_path = output_dir / "75_media_ole_api.pptx"

    # Write tiny WAV to a temp file so add_audio() (file-path API) can read it.

    with tempfile.NamedTemporaryFile(suffix=".wav", delete=False) as tmp:
        tmp.write(_TINY_WAV)
        wav_path = Path(tmp.name)

    try:
        with Presentation.new("Media & OLE API Demo") as prs:
            _add_overview_slide(prs)
            _add_audio_slide(prs, wav_path)
            _add_video_slide(prs)
            _add_ole_reference_slide(prs)

            prs.save(str(output_path))
            print(f"Saved: {output_path}")
    finally:
        wav_path.unlink(missing_ok=True)

    print("\n=== SUMMARY ===")
    print("Demonstrated: add_audio (from bytes via temp file),")
    print("  add_video (fixture-conditional), OLE concept reference slide")


if __name__ == "__main__":
    main()
